specifyweb/context/app_resource.py

```diff
@@ -125,46 +125,53 @@ def get_app_resource_from_db(collection, user, level, resource_name):
-        return (resource.data, resource.spappresource.mimetype, resource.spappresource.id)
+        return resource.data, resource.spappresource.mimetype, resource.spappresource.id
     except Spappresourcedata.MultipleObjectsReturned:
         logger.warning('found multiple appresources for %s', filters)
         resource = Spappresourcedata.objects.filter(**filters)[0]
-        return (resource.data, resource.spappresource.mimetype, resource.spappresource.id)
+        return resource.data, resource.spappresource.mimetype, resource.spappresource.id
     except Spappresourcedata.DoesNotExist:
         # The resource does not exist in the database at the given level.
         return None
 
+# Defines which fields to ignore when looking up
+class Ignore(object):
+    pass
+
 def get_app_resource_dirs_for_level(collection, user, level):
     """Returns a queryset of SpAppResourceDir that match the user/collection context
     at the given level of the hierarchy. In principle the queryset should represent
     a single row. The queryset is returned so that the django ORM can use it to
     build a IN clause when selecting the actual SpAppResource row, resulting in
     a single query to get the resource."""
-    usertype = get_usertype(user)
-    discipline = collection.discipline
+
+    usertype = get_usertype(user) if user is not None else None
+    discipline = collection.discipline if collection is not None else None
 
     # Define what the filters (WHERE clause) are for each level.
     filter_levels = {
         'Columns'    : ('ispersonal', 'usertype', 'collection', 'discipline'),
-        'Personal'   : (True        , 'IGNORE'  , collection  , discipline)  ,
+        'Personal'   : (True        , Ignore() , collection  , discipline)  ,
         'UserType'   : (False       , usertype  , collection  , discipline)  ,
         'Collection' : (False       , None      , collection  , discipline)  ,
         'Discipline' : (False       , None      , None        , discipline)  ,
-        'Common'     : (False       , "Common"  , None        , None)}
+        'Common'     : (False       , Ignore()  , None        , None)
+    }
 
     if level not in filter_levels: return []
 
     # Pull out the column names and values for the given level.
     columns, values = filter_levels['Columns'], filter_levels[level]
-    filters = dict(list(zip(columns, values)))
+    raw_filters = dict(list(zip(columns, values)))
 
     # At the user level, add a clause for the user column.
-    if filters['ispersonal']:
+    if raw_filters['ispersonal']:
         if user is None: return []
-        filters['specifyuser'] = user
-        # when filtering by user, usertype is implied
-        # also if a user's type is changed, what would happen?
-        del filters['usertype']
+        raw_filters['specifyuser'] = user
+
+    filters = {key: value
+               for (key, value) in raw_filters.items()
+               if not isinstance(value, Ignore)}
```

Comment:
don't know about Python, but in JavaScript, instanceof is relatively expensive, and this function is in the hot path (called several times on every page load)

could we do something like this instead?
```py
ignore = {}

# and later:

if value != ignore

```

Reply:
Didn't know js did that!

Fixed in https://github.com/specify/specify7/commit/268ae03e44b0e2923cfb9c340ccdda4c9535ce26